taiga_contrib_ldap_auth_ext/connector.py

```diff
@@ -91,58 +97,66 @@ def login(username: str, password: str) -> tuple:
-                   check_names=True, authentication=SIMPLE,
-                   user=dn, password=password)
-    except Exception as e:
-        error = "LDAP bind failed: %s" % e
-        raise LDAPUserLoginError({"error_message": error})
-
-    # LDAP binding successful, but some values might have changed, or
-    # this is the user's first login, so return them
-    return (username, email, full_name)
+
+    for ldap_value in server_ldap_dict:
+        if server_ldap_dict[ldap_value] is None:
+            continue
+        try:
+            cur_connection = server_ldap_dict[ldap_value]
+            cur_connection.search(search_base=SEARCH_BASE,
+                                                search_filter=search_filter,
+                                                search_scope=SUBTREE,
+                                                attributes=[USERNAME_ATTRIBUTE,
+                                                            EMAIL_ATTRIBUTE, FULL_NAME_ATTRIBUTE],
+                                                paged_size=5)
+        except Exception as e:
+            error = "LDAP login incorrect: %s" % e
+            raise LDAPUserLoginError({"error_message": error})
+
+        # we are only interested in user objects in the response
+        cur_connection.response = [r for r in cur_connection.response if 'raw_attributes' in r and 'dn' in r]
+        # stop if no search results
+        if not cur_connection.response:
+            raise LDAPUserLoginError({"error_message": "LDAP login not found"})
+
+        # handle multiple matches
+        if len(cur_connection.response) > 1:
+            raise LDAPUserLoginError(
+                {"error_message": "LDAP login could not be determined."})
+
+        # handle missing mandatory attributes
+        raw_attributes = cur_connection.response[0].get('raw_attributes')
+        if raw_attributes.get(USERNAME_ATTRIBUTE) or raw_attributes.get(EMAIL_ATTRIBUTE) or raw_attributes.get(
+                FULL_NAME_ATTRIBUTE):
+            raise LDAPUserLoginError({"error_message": "LDAP login is invalid."})
+
+        # attempt LDAP bind
+        username = raw_attributes.get(USERNAME_ATTRIBUTE)[0].decode('utf-8')
+        email = raw_attributes.get(EMAIL_ATTRIBUTE)[0].decode('utf-8')
+        full_name = raw_attributes.get(FULL_NAME_ATTRIBUTE)[0].decode('utf-8')
+        try:
+            dn = str(bytes(cur_connection.response[0].get('dn'), 'utf-8'), encoding='utf-8')
+            Connection(ldap_value, auto_bind=auto_bind, client_strategy=SYNC,
+                       check_names=True, authentication=SIMPLE,
+                       user=dn, password=password)
+        except Exception as e:
+            error = "LDAP bind failed: %s" % e
+            raise LDAPUserLoginError({"error_message": error})
+
+        # LDAP binding successful, but some values might have changed, or
+        # this is the user's first login, so return them
+        return username, email, full_name
```

Comment:
@AminovE99 There is no fallback mechanism here.

If we have for instance 3 LDAP servers which we will call `A` / `B` / `C`, when a user which is in LDAP `B` tries to connect, the feature must follow this process:

1. attempt login to `A` --> user not found or exception raised, then move to next server
1. attempt login to `B` --> user found. return user and proceed with authentication

The current code will crash at login to LDAP `A` and never move to LDAP `B`.

To give another scenario, if we have the same 3 LDAP servers, when a user which is not in any LDAP server tries to connect, the feature must follow this process:

1. attempt login to `A` --> user not found or exception raised, then move to next server
1. attempt login to `B` --> user not found or exception raised, then move to next server
1. attempt login to `C` --> user not found or exception raised, then move to next server
1. no other LDAP available, raise an exception and let `services.py` handle this exception (local fallback)

Given the expected behavior, it might be better to:

- [ ] put the code responsible to connect to one LDAP server into a separate function
- [ ] call this function in loop
- [ ] catch and log all exceptions raised by this function
- [ ] stop the loop and return if successful authentication to LDAP server
- [ ] raise an exception if no successful authentication after going through all the LDAP servers